mitmproxy/command.py

```diff
@@ -159,23 +163,29 @@ def parse_partial(
         """
             Parse a possibly partial command. Return a sequence of ParseResults and a sequence of remainder type help items.
         """
-        buf = io.StringIO(cmdstr)
         parts: typing.List[str] = []
-        lex = lexer(buf)
-        while 1:
-            remainder = cmdstr[buf.tell():]
-            try:
-                t = lex.get_token()
-            except ValueError:
-                parts.append(remainder)
-                break
-            if not t:
-                break
-            parts.append(t)
+
+        rex = pyparsing.QuotedString("\"", escChar='\\', unquoteResults=False) |\
+            pyparsing.QuotedString("'", escChar='\\', unquoteResults=False) |\
+            pyparsing.Combine(pyparsing.Literal('"') + pyparsing.Word(pyparsing.printables + " ") + pyparsing.StringEnd()) |\
+            pyparsing.Word(pyparsing.printables) |\
+            pyparsing.Word(' ')
+
+        rex = rex.copy().leaveWhitespace()
+
+        for t, start, end in rex.scanString(cmdstr):
+            parts.append(t[0])
+
         if not parts:
-            parts = [""]
-        elif cmdstr.endswith(" "):
-            parts.append("")
+            parts = []
+
+        # First item in parts has always to be the command
+        # so we remove any blank tokens from the start of it
+        while True:
+            if parts and parts[0].strip() == '':
+                del parts[0]
+            else:
+                break
```

Comment:
What's the reasoning for special-casing whitespace at the start? It doesn't make much sense, but I don't think we need a special case for it. Am I missing something?

(if I am indeed missing something, can there be more than one blank token at the start? If not, this should become a simple if)

Reply:
In my opinion there shouldn't be any situations where this happens (I changed the code to prevent a whitespace from being added there from the commander), but there was a test case for it with various line brakes or spaces before the actual command. 

If you feed the parser with `"\n \n \n p1"` the response will be `[' ', ' ', ' ', 'p1'], so there is a chance to have more than one blank token prior to the command making the loop necessary.

Reply:
Ok, two thoughts here:

1) Let's fix the parser to return `\n \n \n ` as a single whitespace thing. That seems more correct.
2) What happens if we allow starting spaces in the commander? Can't we treat all spaces equally at a later point and not special case anything before? Of course you are right that spaces at the start don't make much sense, but it shouldn't be harmful and maybe we can reduce complexity if we tolerate it? :)

Reply:
1 - Done
2 - This is where I am now. If we allow whitespaces before the command, then we need to change `CommandManager.execute` to find the first value in the `params` array that has a type of `mitmproxy.types.Cmd` to pass it as the first param to `CommandManager.call_strings`. It is not too pretty, but has to be done.

In that note, we should turn `CommandManager.call_strings` and `CommandManager.call` into private methods. Currently there are 3 entry points available to execute a command, and I don't think that is a good idea. Anything that wants to execute a command, should just use `CommandManager.execute`. If you agree, I can make this change as well.

Reply:
Alright, I made changes to make `CommandManager.execute` the only entry point for command execution and changed `call_strings` to a private method by prepending it with an underscore (it is no _call_strings).

The `execute` method does the appropriate cleaning prior to executing commands.

Hopefully with these changes things are in good shape. If you disagree with anything, just let me know.